File: brain-move/engine/session.py
import time
import json
from pathlib import Path


class Session:
    def __init__(self, profile, scoring, difficulty, safety):
        self.profile = profile
        self.scoring = scoring
        self.difficulty = difficulty
        self.safety = safety
        self.session_id = time.strftime("%Y-%m-%d-%H%M%S")
        self.start_time = None
        self.game_log = []
        self.current_game = None

    def start(self):
        self.start_time = time.time()
        self.safety.reset()

    def log_game(self, game_name, result):
        entry = {
            "session_id": self.session_id,
            "game": game_name,
            "difficulty": self.difficulty.get_level(),
            "result": result,
            "timestamp": time.time(),
        }
        self.game_log.append(entry)

    def end(self):
        duration = time.time() - self.start_time if self.start_time else 0
        summary = {
            "session_id": self.session_id,
            "duration_seconds": round(duration, 1),
            "total_score": self.scoring.get_total(),
            "final_difficulty": self.difficulty.get_level(),
            "safety_events": self.safety.get_events(),
            "games": self.game_log,
        }
        self._save(summary)
        return summary

    def _save(self, summary):
        data_dir = Path(__file__).parent.parent / "data"
        data_dir.mkdir(exist_ok=True)
        path = data_dir / f"session_{self.session_id}.json"
        with open(path, "w") as f:
            json.dump(summary, f, indent=2)

    def get_elapsed_minutes(self):
        if self.start_time is None:
            return 0
        return (time.time() - self.start_time) / 60.0

File: brain-move-web/src/engine/session.ts
import { Scoring } from './scoring'
import { Difficulty } from './difficulty'
import { SafetyEngine } from './safety'
import { Profile } from './profile'
import type { GameResult } from '../types'

export class Session {
  sessionId: string
  startTime: number | null = null
  gameLog: Array<{
    sessionId: string
    game: string
    difficulty: number
    result: GameResult
    timestamp: number
  }> = []
  currentGame: string | null = null

  constructor(
    public profile: Profile,
    public scoring: Scoring,
    public difficulty: Difficulty,
    public safety: SafetyEngine,
  ) {
    this.sessionId = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19)
  }

  start(): void {
    this.startTime = Date.now()
    this.safety.reset()
  }

  logGame(gameName: string, result: GameResult): void {
    this.gameLog.push({
      sessionId: this.sessionId,
      game: gameName,
      difficulty: this.difficulty.getLevel(),
      result,
      timestamp: Date.now(),
    })
  }

  end(): {
    sessionId: string
    durationSeconds: number
    totalScore: number
    finalDifficulty: number
    safetyEvents: Array<{ type: string; details: string }>
    games: Array<{ sessionId: string; game: string; difficulty: number; result: GameResult; timestamp: number }>
  } {
    const duration = this.startTime ? (Date.now() - this.startTime) / 1000 : 0
    const summary = {
      sessionId: this.sessionId,
      durationSeconds: Math.round(duration * 10) / 10,
      totalScore: this.scoring.getTotal(),
      finalDifficulty: this.difficulty.getLevel(),
      safetyEvents: this.safety.getEvents(),
      games: this.gameLog,
    }
    this._save(summary)
    return summary
  }

  private _save(summary: unknown): void {
    try {
      const stored = JSON.parse(localStorage.getItem('brainmove_sessions') ?? '[]')
      stored.push(summary)
      localStorage.setItem('brainmove_sessions', JSON.stringify(stored))
    } catch { /* storage full or unavailable */ }
  }

  getElapsedMinutes(): number {
    if (!this.startTime) return 0
    return (Date.now() - this.startTime) / 60000
  }
}
